onnxruntime/contrib_ops/cuda/collective/nccl_kernels.cc

```diff
@@ -35,21 +42,137 @@ static ncclDataType_t GetNcclDataType(onnxruntime::MLDataType type) {
+
+  FLLOG << "Listening on port " << PortNumber << " for the other GPU processores...\n";
+  word_size--;  // rank 0 is not in word_size
+  while (word_size-- > 0) {
+    int client_fd = accept(fd, nullptr, nullptr); /* accept blocks */
+    if (client_fd < 0) {
+      FLLOGERRNO << ("accept\n"); /* terminate */
+      return -1;
+    }
+    FLLOG << ("Accepted new GPU\n");
+    if (write(client_fd, (nccl_id), sizeof(ncclUniqueId)) != sizeof(ncclUniqueId)) {
+      FLLOGERRNO << ("write\n"); /* terminate */
+      return -1;
+    }
+    close(client_fd);
+  }
+  close(fd);
+  return 0;
+}
+
+
+int ReadFromRank0(ncclUniqueId* nccl_id) {
+  int sockfd = socket(AF_INET,     /* versus AF_LOCAL */
+                      SOCK_STREAM, /* reliable, bidirectional */
+                      0);          /* system picks protocol (TCP) */
+  if (sockfd < 0) {
+    FLLOGERRNO << ("socket");
+    return -1;
+  }
+
+  /* connect to the server: configure server's address 1st */
+  std::string rank0_ip = ParseEnvironmentVariableWithDefault<std::string>("RANK0_IP", "127.0.0.1");
+  int32_t PortNumber = ParseEnvironmentVariableWithDefault<int32_t>("RANK0_PORT", 18888);
+
+  struct sockaddr_in saddr;
+  memset(&saddr, 0, sizeof(saddr));
+  saddr.sin_family = AF_INET;
+  saddr.sin_addr.s_addr = inet_addr(rank0_ip.c_str());
+  saddr.sin_port = htons(PortNumber); /* port number in big-endian */
+  time_t start_time = time(0);
+  int conn_ret = connect(sockfd, (struct sockaddr*)&saddr, sizeof(saddr));
+  while (time(0) - start_time < 40 && conn_ret < 0) {
+    FLLOGERRNO << (" waiting the RANK 0 ready..."); /* terminate */
+    sleep(1);
+    conn_ret = connect(sockfd, (struct sockaddr*)&saddr, sizeof(saddr));
+  }
+  if (conn_ret < 0) {
+    FLLOGERRNO << ("connect"); /* terminate */
+    return -1;
+  }
+
+  if (read(sockfd, (nccl_id), sizeof(ncclUniqueId)) != sizeof(ncclUniqueId)) {
+    FLLOGERRNO << ("read"); /* terminate */
+    return -1;
+  }
+
+  close(sockfd);
+  return 0;
+}
+
+int IPC_Bcast(ncclUniqueId* nccl_id, int rank, int world_size) {
+  if (rank == 0) {
+    if (WriteOnRank0(nccl_id, world_size) != 0) {
+      return (-1);
+    }
+  } else if (ReadFromRank0(nccl_id) != 0) {
+    return (-1);
+  }
+
+  return 0;
+}
+}  // namespace IPC
+
+#endif
+
 #ifdef USE_MPI
-static Status CreateNcclCommByMPI(int world_size, int rank, ncclComm_t* comm) {
+static Status CreateNcclCommunicator(int world_size, int rank, ncclComm_t* comm, bool is_launched_by_mpi) {
   // Create new NCCL communicator
   ncclUniqueId nccl_id;
   if (rank == 0) {
     NCCL_RETURN_IF_ERROR(ncclGetUniqueId(&nccl_id));
   }
-  MPI_CHECK(MPI_Bcast(&nccl_id, sizeof(nccl_id), MPI_BYTE, 0, MPI_COMM_WORLD));
+  if (is_launched_by_mpi) {
+#ifdef USE_MPI
+    MPI_CHECK(MPI_Bcast(&nccl_id, sizeof(nccl_id), MPI_BYTE, 0, MPI_COMM_WORLD));
+#else
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Please compile ORT with USE_MPI.");
+#endif
+  } else if (IPC::IPC_Bcast(&nccl_id, rank, world_size) != 0) {
```

Comment:
IPC is defined within `#ifndef USE_MPI` but used here within `#ifdef USE_MPI` (see line 152).